### PR TITLE
[Model] Add Gemma4 Unified (encoder-free)  support

### PR DESCRIPTION
**Pull request by lucianommartins**
### Summary

Adds `Gemma4UnifiedForConditionalGeneration` support for the encoder-free Gemma 4 12B model family. Unlike the tower-based Gemma 4 variants (E2B, E4B, 26B-A4B, 31B), the unified variant has no vision encoder and no audio encoder: raw pixel patches and audio waveform frames are projected directly into LM space.

### Architecture

- `Gemma4UnifiedForConditionalGeneration` subclasses `Gemma4ForConditionalGeneration` using the `nn.Module.__init__` + full rebuild pattern (same approach as `LightOnOCRForConditionalGeneration`), keeping the parent class free of encoder-free branching
- Vision: `Gemma4UnifiedVisionEmbedder` replaces the vision tower with a `Dense+LayerNorm` pipeline and factorized 2D positional embeddings
- Audio: raw 16kHz waveform frames projected through `Gemma4MultimodalEmbedder` with no conformer audio tower
- `Gemma4UnifiedProcessingInfo` subclasses `Gemma4ProcessingInfo` to handle the `num_soft_tokens` vs `default_output_length` field mapping
- All modalities (image, video, audio) are supported

### Changes

- **New file**: `vllm/model_executor/models/gemma4_unified.py` (466 lines) with `Gemma4UnifiedVisionEmbedder`, `Gemma4UnifiedProcessingInfo`, and `Gemma4UnifiedForConditionalGeneration`
- **gemma4_mm.py**: SigLIP-style quantization check for vision/audio towers (BNB/torchao always quantize, others check dimensions % 64); `suppress_tokens` from `generation_config` applied in `compute_logits`
- **gemma4_mtp.py**: `attention_k_eq_v` + `num_global_key_value_heads` handling for full-attention MTP decoder layers; `suppress_tokens` applied in drafter `compute_logits`; `gemma4_unified_assistant` mapped to `Gemma4MTPModel`
- **gpu_model_runner.py**: audio modality skip for bidirectional attention ranges
- **Registry/config**: `Gemma4UnifiedForConditionalGeneration` registered in model registry, config map, arch config convertor, and spec decode proposer
- **Docs**: updated `supported_models.md` and `mtp.md` with unified variant and 12B assistant model

### Duplicate-work check

No existing PR addresses Gemma 4 Unified / encoder-free support. This is the initial implementation of `Gemma4UnifiedForConditionalGeneration` in vLLM.

### Test commands and results

```bash
# Processing tests (unified image schema, soft-token caps, prompt updates, limit_mm_per_prompt)
python -m pytest tests/models/multimodal/processing/test_gemma4_unified.py -v

# Lint
ruff check vllm/model_executor/models/gemma4_unified.py vllm/model_executor/models/gemma4_mm.py vllm/model_executor/models/gemma4_mtp.py
ruff format --check vllm/model_executor/models/gemma4_unified.py vllm/model_executor/models/gemma4_mm.py vllm/model_executor/models/gemma4_mtp.py

### CODE REVIEW

File: tests/models/registry.py [hunk 1]
```diff
@@ -917,6 +917,12 @@ def check_available_online(
         "google/gemma-4-E2B-it",
         min_transformers_version="5.5.0",
     ),
+    # TODO: update min_transformers_version when Gemma4 Unified lands in
+    # a stable transformers release.
+    "Gemma4UnifiedForConditionalGeneration": _HfExamplesInfo(
+        "google/gemma-4-12B-it",
+        min_transformers_version="5.8.0",
+    ),
```
Comment (mgoin) [inline tests/models/registry.py:926]: Make sure to update before landing or block on `is_available_online=False`

File: vllm/model_executor/models/gemma4.py [hunk 2]
```diff
@@ -1051,11 +1051,14 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
-        # Embedding scale = sqrt(hidden_size)
-        # Downcast to model dtype (bfloat16 etc.) for numerical parity
+        # Embedding scale = sqrt(hidden_size), cast to model dtype to avoid
+        # mixed-precision drift from bf16 * fp32 across deep stacks.
         self.register_buffer(
             "normalizer",
-            torch.tensor(config.hidden_size**0.5),
+            torch.tensor(
+                config.hidden_size**0.5,
+                dtype=self.embed_tokens.weight.dtype,
```
Comment (mgoin) [inline vllm/model_executor/models/gemma4.py:1060]: Can you use a different dtype source here? If embed_tokens is quantized, then "weight" won't exist right. Same problem I found in https://github.com/vllm-project/vllm/pull/44340

File: vllm/model_executor/models/gemma4_mm.py [hunk 3]
```diff
@@ -988,18 +1012,35 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
+            "torchao",
+        ]:
+            tower_quant = quant_config
+        else:
+            vision_cfg = config.vision_config
+            quantizable = (
+                vision_cfg.hidden_size % 64 == 0
+                and vision_cfg.intermediate_size % 64 == 0
+            )
+            tower_quant = quant_config if quantizable else None
```
Comment (mgoin) [inline vllm/model_executor/models/gemma4_mm.py:1030]: We should add this padding support to the quant methods rather than specializing specific quant configs like this, this is a kernel limitation not a checkpoint limitation. If the checkpoint is quantized but we say the layers aren't quantized, then the model will be incorrectly loading weights. We should instead mark those layers in the ignore list in the checkpoint format i.e. compressed-tensors

File: vllm/model_executor/models/gemma4_mm.py [hunk 4]
```diff
@@ -350,20 +373,21 @@ def get_audio_repl(
...
-            torch.zeros(audio_len), sampling_rate
+        num_tokens = self._compute_audio_num_tokens(
+            audio_len, sampling_rate, processor.audio_seq_length
         )
         config = self.get_hf_config()
         token_ids = (
             [config.boa_token_id]
             + [processor.audio_token_id] * num_tokens
-            + [config.eoa_token_id]
+            + [getattr(config, "eoa_token_id", None) or config.eoa_token_index]
```
Comment (mgoin) [inline vllm/model_executor/models/gemma4_mm.py]: nit
```suggestion
            + [getattr(config, "eoa_token_id", config.eoa_token_index)]
```

File: vllm/v1/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -2466,6 +2466,8 @@ def _build_attn_group_metadata(
                 image_doc_ranges = []
                 req_state = self.requests[req_id]
                 for mm_feature in req_state.mm_features:
+                    if mm_feature.modality == "audio":
+                        continue
```
Comment (mgoin) [inline vllm/v1/worker/gpu_model_runner.py:2471]: Will this affect the previous gemma4-with-audio behavior? Maybe worth testing E2B on audio quickly

File: vllm/model_executor/models/gemma4_unified.py [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+
+        per_image_features: list[torch.Tensor] = []
+        for pv, pp in zip(pixel_values, pixel_position_ids, strict=True):
+            pv = pv.unsqueeze(0)
+            pp = pp.unsqueeze(0)
+            embedded = self.vision_embedder(pv, pp)
+            projected = self.embed_vision(embedded.to(target_dtype))
+            padding_mask = (pp.squeeze(0) == -1).all(dim=-1)
+            valid_features = projected.squeeze(0)[~padding_mask]
+            per_image_features.append(valid_features)
```
Comment (mgoin) [inline vllm/model_executor/models/gemma4_unified.py:376]: perf nit: for both image and video we are looping one image/frame at a time through the vision embedder, one forward per item. we should consider batching in the future

File: vllm/model_executor/models/gemma4_unified.py [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+
+        # ---- Encoder-free audio embedder ----
+        self.embed_audio = (
+            Gemma4MultimodalEmbedder(
+                config.audio_config,
+                config.text_config,
+            )
+            if config.audio_config is not None
+            else None
+        )
```
Comment (mgoin) [inline vllm/model_executor/models/gemma4_unified.py:291]: These are constructed without `quant_config` where it is passed through in gemma4_mm, so it could be supported the same way I think. If intentional should note

File: vllm/model_executor/models/gemma4_mm.py (hunk 4 above)
Comment (lucianommartins) [reply vllm/model_executor/models/gemma4_mm.py]: done!

File: tests/models/registry.py (hunk 1 above)
Comment (lucianommartins) [reply tests/models/registry.py:926]: done!

**Review by mgoin — APPROVED** [review]

**Comment by lucianommartins** [comment]
hey @mgoin @ywang96 @zifeitong - the only failed test was the Llama TP+SP (sequence parallelism) test with Ray -- unrelated to this PR.

can we go ahead and merge it?

**Comment by ThomasKluiters** [comment]
When audio requests are batched together I obtain an error. input_features_padded is a list, but                
  _process_audio_input() assumes a tensor and calls .squeeze(1) though, this has been an issue with previous gemma versions as well
